### PR TITLE
Support synthetic source together with ignore_malformed in histogram fields

### PR DESCRIPTION
**Pull request by lkts**
Contributes to #106483.

### CODE REVIEW

File: server/src/main/java/org/elasticsearch/index/mapper/CompositeSyntheticFieldLoader.java [hunk 1]
```diff
@@ -0,0 +1,174 @@
...
+
+import static java.util.Collections.emptyList;
+
+/**
+ * A {@link SourceLoader.SyntheticFieldLoader} that uses a set of sub-loaders
+ * to produce synthetic source for the field.
+ * Typical use case is to gather field values from doc_values and append malformed values
+ * stored in a different field in case of ignore_malformed being enabled.
+ */
+public class CompositeSyntheticFieldLoader implements SourceLoader.SyntheticFieldLoader {
```
Comment (lkts) [inline server/src/main/java/org/elasticsearch/index/mapper/CompositeSyntheticFieldLoader.java:30]: I noticed after implementing this that this is very close to what `ObjectMapper.SyntheticSourceFieldLoader` does. Maybe we can unify some code later.

This is also an alternative approach to current implementation of f.e. `SortedNumericDocValuesSyntheticFieldLoader` where malformed values handling is implemented explicitly. That logic is repeated in multiple loaders that handle different doc values types. I obviously didn't refactor that in this PR but wanted to gather some thoughts.

**Comment by lkts** [comment]
@elasticmachine update branch

File: server/src/main/java/org/elasticsearch/index/mapper/CompositeSyntheticFieldLoader.java (hunk 1 above)
Comment (martijnvg) [reply server/src/main/java/org/elasticsearch/index/mapper/CompositeSyntheticFieldLoader.java:30]: +1 I think in a followup we can explore how to have a common base class for this class and `ObjectMapper.SyntheticSourceFieldLoader`.

**Review by martijnvg — APPROVED** [review]
LGTM

File: server/src/main/java/org/elasticsearch/index/mapper/CompositeSyntheticFieldLoader.java [hunk 2]
```diff
@@ -0,0 +1,174 @@
...
+    /**
+     * Layer that loads malformed values stored in a dedicated field with a conventional name.
+     * @see IgnoreMalformedStoredValues
+     */
+    public static class MalformedValuesLayer implements SyntheticFieldLoaderLayer {
+        private final String fieldName;
+        private List<Object> values;
+
+        public MalformedValuesLayer(String fieldName) {
+            this.fieldName = fieldName + "._ignore_malformed";
```
Comment (kkrik-es) [inline server/src/main/java/org/elasticsearch/index/mapper/CompositeSyntheticFieldLoader.java:133]: "._ignore_malformed" should be a const somewhere.

File: server/src/main/java/org/elasticsearch/index/mapper/CompositeSyntheticFieldLoader.java [hunk 3]
```diff
@@ -0,0 +1,174 @@
...
+            return values.isEmpty() == false;
+        }
+
+        @Override
+        public void write(XContentBuilder b) throws IOException {
+            for (Object v : values) {
+                if (v instanceof BytesRef r) {
+                    XContentDataHelper.decodeAndWrite(b, r);
+                } else {
+                    b.value(v);
```
Comment (kkrik-es) [inline server/src/main/java/org/elasticsearch/index/mapper/CompositeSyntheticFieldLoader.java:163]: What's the use case for this one? I thought malformed values are always encoded.

File: x-pack/plugin/analytics/src/main/java/org/elasticsearch/xpack/analytics/mapper/HistogramFieldMapper.java [hunk 4]
```diff
@@ -491,76 +514,85 @@ protected SyntheticSourceMode syntheticSourceMode() {
...
-                while (value.next()) {
-                    b.value(value.count());
-                }
-                b.endArray();
+        @Override
+        public void write(XContentBuilder b) throws IOException {
+            if (binaryValue == null) {
+                return;
+            }
+            b.startObject();
```
Comment (kkrik-es) [inline x-pack/plugin/analytics/src/main/java/org/elasticsearch/xpack/analytics/mapper/HistogramFieldMapper.java:569]: Why was this changed from `b.startObject(simpleName());` ?

File: x-pack/plugin/src/yamlRestTest/resources/rest-api-spec/test/analytics/histogram.yml [hunk 5]
```diff
@@ -306,3 +306,54 @@ histogram with large count values:
...
+      _source:
+        latency: "quick brown fox"
+
+  - do:
+      get:
+        index: histo_synthetic
+        id: 2
+  - match:
+      _source:
+        latency: [{"values": [2.0], "counts": [2]}, {"values": [1.0], "counts": [1], "hello": "world"}, 123, 456, "fox"]
```
Comment (kkrik-es) [inline x-pack/plugin/src/yamlRestTest/resources/rest-api-spec/test/analytics/histogram.yml:359]: We miss that we got `[123, 456]` as a pair.. Not a biggie, wonder if there's an easy way to catch the array.

**Review by kkrik-es — APPROVED** [review]
Nice, just a few minor ones.

File: x-pack/plugin/src/yamlRestTest/resources/rest-api-spec/test/analytics/histogram.yml (hunk 5 above)
Comment (lkts) [reply x-pack/plugin/src/yamlRestTest/resources/rest-api-spec/test/analytics/histogram.yml:359]: This is intentional, this is how it works everywhere.

File: x-pack/plugin/analytics/src/main/java/org/elasticsearch/xpack/analytics/mapper/HistogramFieldMapper.java (hunk 4 above)
Comment (lkts) [reply x-pack/plugin/analytics/src/main/java/org/elasticsearch/xpack/analytics/mapper/HistogramFieldMapper.java:569]: This is because composite loader writes that now. It is possible that there are malformed values so this is now not an object but an array.

File: server/src/main/java/org/elasticsearch/index/mapper/CompositeSyntheticFieldLoader.java (hunk 3 above)
Comment (lkts) [reply server/src/main/java/org/elasticsearch/index/mapper/CompositeSyntheticFieldLoader.java:163]: If it's f.e. text we skip encoding in some fields. This is for compatibility with existing code.
